packages/ai-jsx/src/lib/openai.tsx

```diff
@@ -387,23 +409,48 @@ export async function* OpenAIChatModel(
     );
   }
 
-  const openaiFunctions = !props.functionDefinitions
+  // Merge adjacent tool calls.
+  const mergedMessages = messages.reduce<OpenAIClient.Chat.ChatCompletionMessageParam[]>((mergedMessages, message) => {
+    if (message.role === 'assistant' && message.tool_calls && message.content === '') {
+      const lastMessage = mergedMessages[mergedMessages.length - 1];
+      if (lastMessage.role === 'assistant' && lastMessage.tool_calls) {
+        // Merge with the last message.
+        return [
+          ...mergedMessages.slice(0, -1),
+          {
+            ...lastMessage,
+            tool_calls: [...lastMessage.tool_calls, ...message.tool_calls],
+          },
+        ];
+      }
+    }
+    return mergedMessages.concat([message]);
+  }, []);
+
+  const openaiTools = !props.functionDefinitions
     ? []
-    : Object.entries(props.functionDefinitions).map(([functionName, functionDefinition]) => ({
-        name: functionName,
-        description: functionDefinition.description,
-        parameters: getParametersSchema(functionDefinition.parameters),
-      }));
+    : Object.entries(props.functionDefinitions).map<OpenAIClient.Chat.ChatCompletionTool>(
+        ([functionName, functionDefinition]) => ({
+          function: {
+            name: functionName,
+            description: functionDefinition.description,
+            parameters: getParametersSchema(functionDefinition.parameters),
+          },
+          type: 'function',
+        })
+      );
 
   const openai = getContext(openAiClientContext)();
-  const chatCompletionRequest = {
+  const chatCompletionRequest: OpenAIClient.Chat.Completions.ChatCompletionCreateParamsStreaming = {
     model: props.model,
     max_tokens: props.maxTokens,
     temperature: props.temperature,
     top_p: props.topP,
-    messages,
-    functions: openaiFunctions.length > 0 ? openaiFunctions : undefined,
-    function_call: props.forcedFunction ? { name: props.forcedFunction } : undefined,
+    messages: mergedMessages,
+    tools: openaiTools.length > 0 ? openaiTools : undefined,
+    tool_choice: props.forcedFunction
+      ? { function: { name: props.forcedFunction }, type: 'function' as const }
```

Comment:
The idea is that these are `functions` and there can be other `tools` such as Code Interpreter that may be hard coded? Is that what `type` distinguishes?

Reply:
Per their [docs](https://platform.openai.com/docs/api-reference/chat/create), currently only `"function"` is supported. But presumably they're leaving the door open for other types in the future.

Reply:
Seems to me like we can already pass in `{"type": "code_interpreter"}` and `{"type": "retrieval"}` to use OAI-native tools:
![image](https://github.com/fixie-ai/ai-jsx/assets/5062458/913dec62-4033-4900-bad7-2063b1cbe300)
![image](https://github.com/fixie-ai/ai-jsx/assets/5062458/82db633c-5a2e-46b6-9be1-255a495c9c50)